### PR TITLE
Added LuisAdaptiveRecognizer

### LINKED ISSUE #1637: [PORT] External Entities & Dynamic lists
> Port this change from botbuilder-dotnet/master branch:
https://github.com/microsoft/botbuilder-dotnet/pull/3300

Fix for #3156 which now enables external entity recognizers and dynamic lists for declarative LUIS recognizers.
*** Breaking change for declarative LUIS Recognizers ***
They now use expressions so to bind you need to do "=luis.settings.endpointKey" rather than "{luis.settings.endpointKey}".  There is also a PR to update bf-cli.

Dropped $role of lg and memorypath since not used.
Changed it so $role of expression would add string type.
Updated TestBot to include settings.
Moved all of the PropertyExpression stuff to its own library.  Once it no longer depends on .lg we can move it into the actual expression library.

# Changed projects
* Microsoft.Bot.Builder.AI.LUIS
* Microsoft.Bot.Builder.Dialogs.Adaptive
* Microsoft.Bot.Builder.Dialogs.Declarative
* Microsoft.Bot.Expressions.Properties
* Microsoft.Bot.Builder.AI.LUIS.Tests
* Microsoft.Bot.Builder.Ai.LUISV3.tests
* Microsoft.Bot.Builder.Dialogs.Adaptive.Templates.Tests
* Microsoft.Bot.Builder.Dialogs.Adaptive.Tests
* Microsoft.Bot.Builder.LanguageGeneration.Tests
* Microsoft.Bot.Builder.TestBot.Json

[R8,adaptive]

### PR DESCRIPTION
**Pull request by Stevenic**
Fixes #1637

## Description
Added missing LuisAdaptiveRecognizer class

### CODE REVIEW

File: libraries/botbuilder-dialogs-adaptive/src/luis/luisAdaptiveRecognizer.ts [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+     * Key used to talk to a LUIS endpoint.
+     */
+    public endpointKey: StringExpression;
+
+    /**
+     * External entity recognizer.
+     * @remarks
+     * This recognizer is run before calling LUIS and the results are passed to LUIS.
+     */
+    // public externalEntityRecognizer: Recognizer;
```
Comment (stevengum) [inline libraries/botbuilder-dialogs-adaptive/src/luis/luisAdaptiveRecognizer.ts:47]: Should this be uncommented?

File: libraries/botbuilder-dialogs-adaptive/src/luis/luisAdaptiveRecognizer.ts [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+     * LUIS dynamic list.
+     */
+    public dynamicLists: ArrayExpression<any>;
+
+    /**
+     * LUIS endpoint to query.
+     * @remarks
+     * For example: "https://westus.api.cognitive.microsoft.com"
+     */
+    public endpoint: StringExpression;
```
Comment (stevengum) [inline libraries/botbuilder-dialogs-adaptive/src/luis/luisAdaptiveRecognizer.ts:35]: Is this the completed endpoint with the AppId? Or is this the host?

What value is [`ExpressionProperty.prototype.getValue`](https://github.com/microsoft/botbuilder-js/blob/1a2a64a7a0fa2cb9da65f9dfdf88d22fa78a8000/libraries/adaptive-expressions/src/expressionProperties/expressionProperty.ts#L57) supposed to return?

File: libraries/botbuilder-dialogs-adaptive/src/luis/luisAdaptiveRecognizer.ts [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+
+    /**
+     * Telemetry client.
+     */
+    public telemetryClient: BotTelemetryClient;
+
+    public async recognize(dialogContext: DialogContext, activity: Activity) {
+        // Validate passed in activity matches turn activity 
+        const context = dialogContext.context
+        if (activity && (!context.activity == null || context.activity.type != ActivityTypes.Message || context.activity.text != activity.text)) {
```
Comment (stevengum) [inline libraries/botbuilder-dialogs-adaptive/src/luis/luisAdaptiveRecognizer.ts:63]: `!context.activity == null` should never evaluate to true.

I've refactored the code a little in an attempt to improve readability and used strict non-equality checks:

```suggestion
        const context = dialogContext.context;
        const utteranceMatches: boolean = (context.activity !== undefined && context.activity !== null)
            || context.activity.type !== ActivityTypes.Message
            || context.activity.text != activity.text;
        
        if (activity && utteranceMatches) {
```

I left `context.activity.text != activity.text`, but maybe that should be changed as well?

File: libraries/botbuilder-dialogs-adaptive/src/luis/luisAdaptiveRecognizer.ts (hunk 1 above)
Comment (Stevenic) [reply libraries/botbuilder-dialogs-adaptive/src/luis/luisAdaptiveRecognizer.ts:47]: No... We can't support it right now as it requires changes to the base luis recognizer.

File: libraries/botbuilder-dialogs-adaptive/src/luis/luisAdaptiveRecognizer.ts (hunk 2 above)
Comment (Stevenic) [reply libraries/botbuilder-dialogs-adaptive/src/luis/luisAdaptiveRecognizer.ts:35]: I copied it verbatim from C# so I believe its just endpoint URL as in the comments.

File: libraries/botbuilder-dialogs-adaptive/src/luis/luisAdaptiveRecognizer.ts (hunk 3 above)
Comment (Stevenic) [reply libraries/botbuilder-dialogs-adaptive/src/luis/luisAdaptiveRecognizer.ts:63]: Good catch... I forgot to remove the "= null" part

Comment (Stevenic) [reply libraries/botbuilder-dialogs-adaptive/src/luis/luisAdaptiveRecognizer.ts]: Your version had issues as well but the changed version should work

---
In reply to: [418658435](https://github.com/microsoft/botbuilder-js/pull/2149#discussion_r418658435) [](ancestors = 418658435)

**Pull request merged**